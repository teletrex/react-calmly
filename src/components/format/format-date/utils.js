/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */



import moment from 'moment-timezone';

import { DATE_FORMAT } from './presets';

const shortenYear = format => format.replace(/Y+/g, val => val.slice(2));

const shortenDayMonth = format => format.replace(/D+|M+/g, val => val.slice(1));

export const getFormats = (locale, longDateFormat) => {
  const format = moment.localeData(locale).longDateFormat(longDateFormat);
  const reverseFormat = format
    .split('')
    .reverse()
    .join('');
  const shortYearFormat = shortenYear(format);
  const shortDayMonthFormat = shortenDayMonth(format);
  const shortDayMonthYearFormat = shortenYear(shortDayMonthFormat);

  return {
    format,
    reverseFormat,
    shortYearFormat,
    shortDayMonthFormat,
    shortDayMonthYearFormat,
    yearOnly: 'YYYY',
  };
};

const separatorDictironary = {
  Dot: '.',
  Comma: ',',
  ForwardSlash: '/',
  Hyphen: '-',
};

const trimSeparators = format => format.replace(/^[./\-,]+|[./\-,]+$/g, '');
const replaceSeparatorBy = (format, newSeparator) => format.replace(/[.,/-]/g, newSeparator);

export const getAllSeparatorsShortDateFormats = locale => {
  const formats = getFormats(locale, DATE_FORMAT.SHORT_DATE);

  return Object.entries(separatorDictironary).reduce(
    (allFormats, [separatorName, separator]) => {
      Object.entries(formats).forEach(([formatName, format]) => {
        // eslint-disable-next-line no-param-reassign
        allFormats[`${formatName}${separatorName}`] = replaceSeparatorBy(
          trimSeparators(format),
          separator
        );
      });
      return allFormats;
    },
    { ...formats }
  );
};
