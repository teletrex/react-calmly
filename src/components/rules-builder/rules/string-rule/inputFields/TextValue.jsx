/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';

//import TextInput from '../../../../text-input';
import  {useTranslation}  from '../../../../../translation';
import { TextInput } from '@carbon/react';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { PROPTYPES } from '../../constants';
import { getPrefixedClasses } from '../../helpers';

const TextValue = ({ id, onValueChange, operator, placeholder, value }) => {
  const { t } = useTranslation();
  const { isLight, isReadOnly } = useContext(RulesBuilderContext);
  const [currentValue, setCurrentValue] = useState(value);

  useEffect(() => {
    setCurrentValue(value);
  }, [operator, value]);

  const handleBlur = ({ target }) => {
    const { value: newValue } = target || {};

    if (newValue !== value) {
      onValueChange(newValue);
    }
  };

  const handleOnChange = ({ target }) => {
    setCurrentValue(target?.value);
  };

  if (isReadOnly) {
    return (
      <span className={getPrefixedClasses(['--rule-value-readonly', '--text-space-left'])}>
        {value}
      </span>
    );
  }

  return (
    <TextInput
      hideLabel
      id={id}
      labelText={t('String rule')}
      light={isLight}
      name={id}
      onBlur={handleBlur}
      onChange={handleOnChange}
      placeholder={placeholder || t('Enter value')}
      value={currentValue}
    />
  );
};

TextValue.propTypes = {
  id: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
  operator: PROPTYPES.OPERATOR,
  placeholder: PropTypes.string,
  value: PropTypes.string,
};

TextValue.defaultProps = {
  operator: null,
  placeholder: '',
  value: '',
};

export default TextValue;
