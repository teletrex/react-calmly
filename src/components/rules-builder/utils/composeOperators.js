/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import compact from 'lodash/compact';
import map from 'lodash/map';
import isString from 'lodash/isString';
import find from 'lodash/find';

export const composeOperators = (operators, allowedOperators) =>
  compact(
    map(operators, operator => {
      if (isString(operator)) {
        return find(allowedOperators, ({ value }) => value === operator);
      }

      return operator;
    })
  );
