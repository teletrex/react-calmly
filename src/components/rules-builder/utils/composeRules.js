/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import isArray from 'lodash/isArray';
import omit from 'lodash/omit';
import uid from '@carbon/react/es/tools/uniqueId';

const mapRule = (rule, level, oneConjunctionMode) => {
  if (isArray(rule.data)) {
    let operands = 0;
    let maxLevelInside = 0;
    const data = rule.data.map(item => {
      const ret = mapRule(item, level + 1);
      operands += ret.operands;
      maxLevelInside = Math.max(maxLevelInside, ret.maxLevelInside);
      return ret;
    });
    return {
      ...rule,
      id: uid(),
      conjunction: oneConjunctionMode || rule.conjunction,
      level,
      maxLevelInside: maxLevelInside + 1,
      operands,
      data,
    };
  }

  return {
    ...rule,
    id: uid(),
    level,
    maxLevelInside: 0,
    operands: 1,
    data: rule.data,
  };
};

const mapRuleElements = (rule, createRule) => {
  const data = isArray(rule.data)
    ? rule.data.map(item => mapRuleElements(item, createRule))
    : createRule(rule.criteriaKey, rule.data, rule.timeframe);

  return {
    ...omit(rule, 'timeframe'),
    data,
  };
};

const composeRules = (rules, createRule, oneConjunctionMode) =>
  mapRuleElements(mapRule(rules, 0, oneConjunctionMode) || [], createRule);

export default composeRules;
