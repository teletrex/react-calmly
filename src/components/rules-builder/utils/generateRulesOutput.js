/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import filter from 'lodash/filter';
import isArray from 'lodash/isArray';
import isEmpty from 'lodash/isEmpty';
import map from 'lodash/map';
import omitBy from 'lodash/omitBy';
import pick from 'lodash/pick';

const generateRulesOutput = rules => {
  const { conjunction: topLevelConjunction, data: rulesData } = rules;

  const outputData = map(filter(rulesData, 'data'), rule => {
    const { data, conjunction, criteriaKey } = rule;
    const isGroup = isArray(data) && conjunction && !criteriaKey;

    return isGroup
      ? generateRulesOutput(rule)
      : omitBy(
          {
            criteriaKey,
            data: pick(data.ruleData, 'operator', 'value'),
            timeframe: pick(data.ruleTimeframe, 'operator', 'value'),
          },
          isEmpty
        );
  });

  return {
    conjunction: topLevelConjunction,
    data: outputData,
  };
};

export default generateRulesOutput;
