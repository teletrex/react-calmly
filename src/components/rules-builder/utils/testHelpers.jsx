/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

import RulesBuilderContext from '../RulesBuilderContext';
import {
  ERROR_RELEGATION_REASON_TAG,
  relegateErrors,
} from '../../utils/test-helpers/relegate-errors';

export const datePickerErrorWrapper = fn => {
  relegateErrors(
    {
      reasonTag: ERROR_RELEGATION_REASON_TAG.FAILED_TO_RESOLVE,
      reason: "Couldn't trace the 'uncontrolled input error' in date component.",
      expectedErrorCount: 1,
    },
    () => {
      fn();
    }
  );
};

export const renderWithContext = (component, contextValue = {}) =>
  render(
    <RulesBuilderContext.Provider value={contextValue}>{component}</RulesBuilderContext.Provider>
  );
