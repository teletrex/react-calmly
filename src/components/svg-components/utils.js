/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const getSVGSizeProps = (size, height, width) => {
  const props = size ? { width: size, height: size } : {};
  if (height) {
    props.height = height;
  }
  if (width) {
    props.width = width;
  }
  return props;
};
