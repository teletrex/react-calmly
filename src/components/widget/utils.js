/* <LICENSE>
*
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
*
* </LICENSE>
*  */




import { createCssClassBuilder } from '../../utils';
import { sortObjectByValue } from '../../utils/ObjectUtils';

import { breakpoints } from './constants';

export const widgetClassBuilder = createCssClassBuilder('widget');

export const getBreakpointFromWidth = width => {
  const sortedBreakpoints = sortObjectByValue(breakpoints);

  const lowerBP = sortedBreakpoints[0][0];

  return sortedBreakpoints.reduce((currentBP, [targetBP, BPWidth]) => {
    if (width > BPWidth) {
      return targetBP;
    }
    return currentBP;
  }, lowerBP);
};

export const getGetNewPosition = layout => {
  const xByRow = { 0: 0 };
  let lowestRow = 0;

  layout.forEach(item => {
    const nextAvailableXRelativeToItem = item.x + item.w;
    const lowestItemRow = item.y + item.h - 1;

    // the item can occupy more than one row
    // save his x to each of those rows
    for (let l = 0; l < item.h; l++) {
      const currentRow = l + item.y;
      const rowX = xByRow[currentRow] || 0;
      if (nextAvailableXRelativeToItem > rowX) {
        xByRow[currentRow] = nextAvailableXRelativeToItem;
      }
    }

    if (lowestItemRow > lowestRow) {
      lowestRow = lowestItemRow;
    }
  });

  const proposedX = xByRow[lowestRow];

  return {
    // since we have 6 columns and min width is 2,
    // we cannot place an element after the 4th column,
    // we do, instead, put it on 0 becase it will land on beggining of the next line.
    x: proposedX > 4 ? 0 : proposedX,
    // always put it one line below because even if we don't need to
    // break line, react-grid-layout will vertically compress.
    y: lowestRow + 1,
  };
};
