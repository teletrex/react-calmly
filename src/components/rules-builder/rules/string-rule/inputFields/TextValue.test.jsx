/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { renderWithContext } from '../../../utils/testHelpers';

import TextValue from './TextValue';

const initialValue = 'abc';

const defaultProps = {
  id: 'TestStringRuleElement',
  onValueChange: jest.fn(),
};

describe('TextValue', () => {
  beforeEach(() => {
    defaultProps.onValueChange.mockReset();
  });

  it('should render correctly', () => {
    const { container } = render(<TextValue {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a provided value', () => {
    const { container, getByDisplayValue } = render(
      <TextValue {...defaultProps} value={initialValue} />
    );
    expect(container).toMatchSnapshot();
    expect(getByDisplayValue(initialValue)).toBeInTheDocument();
  });

  it('should render correctly for read-only', () => {
    const { container, getByText } = renderWithContext(
      <TextValue {...defaultProps} value={initialValue} />,
      { isReadOnly: true }
    );
    expect(container).toMatchSnapshot();
    expect(getByText(initialValue)).toBeInTheDocument();
  });

  it('should set the rule value when a value is entered', () => {
    const { getByPlaceholderText } = render(<TextValue {...defaultProps} />);
    const textInput = getByPlaceholderText('Enter value');

    fireEvent.change(textInput, { target: { value: initialValue } });
    fireEvent.blur(textInput);

    expect(defaultProps.onValueChange).toHaveBeenCalledWith(initialValue);
  });

  it('should not update the rule if the value was not changed', () => {
    const { getByPlaceholderText } = render(<TextValue {...defaultProps} value={initialValue} />);
    const textInput = getByPlaceholderText('Enter value');

    fireEvent.blur(textInput);

    expect(defaultProps.onValueChange).toHaveBeenCalledTimes(0);
  });
});
