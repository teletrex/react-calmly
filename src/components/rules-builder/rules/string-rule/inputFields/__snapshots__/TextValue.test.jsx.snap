// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextValue should render correctly 1`] = `
<div>
  <div
    class="bx--form-item bx--text-input-wrapper"
  >
    <span
      class="bx--text-input__label-wrapper bx--visually-hidden"
    >
      <label
        class="bx--label"
        for="TestStringRuleElement"
      >
        String rule
      </label>
      
    </span>
    <div
      class="bx--text-input__field-wrapper"
    >
      <input
        class="bx--text-input"
        id="TestStringRuleElement"
        name="TestStringRuleElement"
        placeholder="Enter value"
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`TextValue should render correctly for read-only 1`] = `
<div>
  <span
    class="bx--rule-value-readonly bx--text-space-left"
  >
    abc
  </span>
</div>
`;

exports[`TextValue should render correctly with a provided value 1`] = `
<div>
  <div
    class="bx--form-item bx--text-input-wrapper"
  >
    <span
      class="bx--text-input__label-wrapper bx--visually-hidden"
    >
      <label
        class="bx--label"
        for="TestStringRuleElement"
      >
        String rule
      </label>
      
    </span>
    <div
      class="bx--text-input__field-wrapper"
    >
      <input
        class="bx--text-input"
        id="TestStringRuleElement"
        name="TestStringRuleElement"
        placeholder="Enter value"
        type="text"
        value="abc"
      />
    </div>
  </div>
</div>
`;
